crates/wasi-common/src/sys/unix/host_impl.rs

```diff
@@ -158,20 +158,14 @@ pub(crate) fn nix_from_oflags(oflags: wasi::__wasi_oflags_t) -> OFlag {
 }
 
 pub(crate) fn filetype_from_nix(sflags: SFlag) -> FileType {
-    if sflags.contains(SFlag::IFCHR) {
-        FileType::CharacterDevice
-    } else if sflags.contains(SFlag::IFBLK) {
-        FileType::BlockDevice
-    } else if sflags.contains(SFlag::IFSOCK) {
-        FileType::SocketStream
-    } else if sflags.contains(SFlag::IFDIR) {
-        FileType::Directory
-    } else if sflags.contains(SFlag::IFREG) {
-        FileType::RegularFile
-    } else if sflags.contains(SFlag::IFLNK) {
-        FileType::Symlink
-    } else {
-        FileType::Unknown
+    match sflags {
```

Comment:
This is a workaround for a `yanix` bug: `SFlag` should not be a bit field! It corresponds to `stat.st_mode` file types, and use as a bitfield results in situations such as a symlink (`0o12`) being also reported as a regular file (`0o10`) _and_ a character device (`0o02`). Because of the lookup order here, when I adjusted `path_get` to use wasi file types, symlinks file types were being changed to character devices :( There will be a followup PR to fix this in yanix.

Reply:
That's an interesting one, thanks for catching it! I think though that `SFlag` is correctly a bit field as prescribed by POSIX (see [inode(7)]). However, is it that we were using it wrong all along? POSIX seems to prescribe that we should do bit masking with `SFlag::S_IFMT` to get the filetype.

[inode(7)]: http://man7.org/linux/man-pages/man7/inode.7.html

Reply:
I think the _mask_ is to pick the four bits that comprise the enumeration - note that the example is `if ((sb.st_mode & S_IFMT) == S_IFREG) {` rather than `if ((sb.st_mode & S_IFREG) == S_IFREG) {`.

Also, if it's a bit mask, I don't think there would be a way to disambiguate
```
           S_IFSOCK   0140000   socket
           S_IFLNK    0120000   symbolic link
```
in the face of `S_IFREG`, `S_IFDIR`, and `S_IFCHR` overlapping with one of those bits?

I think `SFlag` ended up not used much in wasi-common, I didn't see other use sites with a quick look when I was changing `path_get`, so maybe no one was "using" it wrong in the first place? :grin:

Reply:
Right, exactly, the `S_IFMT` is to pick the bits that comprise the filetype enumeration in `st_mode` while ignoring the mode and permission bits. I guess the naming is off here. `st_mode` *is* the bit field, whereas we should have used an enumeration type like `#[repr(u16)] SFiletype` or whatnot to represent *only* the filetype info, remembering though *not* to include `S_IFMT` as part of the enumeration, but rather use it to construct the values. Would you agree here? The interesting thing though is that the same inaccuracy is present in [nix-rust] and I wonder if it's actually used as intended there :thinking:.

[nix-rust]: https://github.com/nix-rust/nix/blob/54cd6fe77977652256d695d565e1b1a5fe4d5613/src/sys/stat.rs#L13

Reply:
I'm really glad you found this bug actually, since as it turns out, when converting between `SFlags` and `host::FileType` (in wasi-common), we don't handle the case of the (named) pipe (`S_IFIFO`).

I've now drafted out a rewrite of this and would really appreciate your feedback. The PR: #. I've also cc'd you in the PR itself. Also, it'd be great if you could verify that the proposed fix the bug you've discovered.

Reply:
I've given it a quick read and it looks good! I'm pretty sure it fixes this, though I'll take a closer look Monday or Tuesday.

Reply:
Agh, I neglected to update here - yes, this is fixed with #945!